File: transport/transport.go
// Package transport turns the `tls` fragment of a service's configuration
// into the two TLS configurations a service needs: one for the listener it
// serves, one for the connections it makes.
//
// It does three things, and deliberately no more:
//
//  1. loads the certificate the platform mounted, and RELOADS it when the
//     platform replaces it;
//  2. presents it, as a server and as a client;
//  3. admits a peer by the ACCOUNT it runs as, read from the identity in its
//     certificate, against a list the configuration gives.
//
// It does not fetch a certificate, mint one, or talk to an authority. That is
// the platform's job, and a service that did it would be asserting an identity
// rather than presenting one it was given.
package transport

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"log/slog"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"
)

// Mode says which listeners a service serves.
type Mode string

const (
	// Off is cleartext only. It is the default, and it is what a service
	// installed on a platform that provides no identity must fall back to.
	Off Mode = "off"

	// Permissive serves both, on two ports, so one edge can migrate at a
	// time. It is a migration state and should carry a date.
	Permissive Mode = "permissive"

	// Strict serves only the authenticated listener.
	Strict Mode = "strict"
)

// Peer is an account that may call this service.
type Peer struct {
	Namespace      string `json:"namespace"`
	ServiceAccount string `json:"serviceAccount"`
}

// Config is the `tls` fragment, decoded.
type Config struct {
	Mode        Mode   `json:"mode"`
	Address     string `json:"address"`
	CertFile    string `json:"certFile"`
	KeyFile     string `json:"keyFile"`
	CAFile      string `json:"caFile"`
	TrustDomain string `json:"trustDomain"`
	Peers       []Peer `json:"peers"`
}

// Identity is a loaded workload identity.
type Identity struct {
	cfg   Config
	roots *x509.CertPool
	log   *slog.Logger

	// The certificate, and what it was loaded from. A rotation replaces the
	// files in place, so the modification time is what says it is stale.
	mu       sync.RWMutex
	cert     *tls.Certificate
	loadedAt time.Time
}

// Load reads the mounted identity. It returns nil when the mode is off,
// which callers treat as "serve cleartext" rather than as an error: a chart
// whose default is off must produce a service that runs.
//
// The logger is where REFUSALS go, and passing one is not optional in
// spirit. A caller is told only that it was refused — telling it which rule
// rejected it describes the allow-list to whoever is probing — so the
// server's log is the only place the reason exists. A service that refuses
// everyone for an unrelated reason and a service that is working look
// identical without it. A nil logger is accepted for a test and discards
// them.
func Load(cfg Config, log *slog.Logger) (*Identity, error) {
	if cfg.Mode == "" {
		cfg.Mode = Off
	}

	switch cfg.Mode {
	case Off:
		return nil, nil
	case Permissive, Strict:
	default:
		return nil, fmt.Errorf("transport mode %q is not off, permissive or strict", cfg.Mode)
	}

	if cfg.Mode == Permissive && cfg.Address == "" {
		return nil, fmt.Errorf("permissive needs an address for the authenticated listener: one listener cannot be both")
	}

	if cfg.TrustDomain == "" {
		return nil, fmt.Errorf("no trust domain, so every peer's identity would be admitted from anywhere")
	}

	pem, err := os.ReadFile(cfg.CAFile)
	if err != nil {
		return nil, fmt.Errorf("read the trust bundle: %w", err)
	}

	roots := x509.NewCertPool()
	if !roots.AppendCertsFromPEM(pem) {
		return nil, fmt.Errorf("the trust bundle at %s holds no certificate", cfg.CAFile)
	}

	if log == nil {
		log = slog.New(slog.DiscardHandler)
	}

	id := &Identity{cfg: cfg, roots: roots, log: log}
	if _, err := id.certificate(); err != nil {
		return nil, err
	}

	return id, nil
}

// Mode is what the configuration asked for; Off when nothing was loaded.
func (i *Identity) Mode() Mode {
	if i == nil {
		return Off
	}

	return i.cfg.Mode
}

// certificate returns the current certificate, re-reading it when the files
// on disk have changed.
//
// Re-reading rather than holding what start-up loaded is the whole point. A
// platform rotates these on an hour or less; a process that cached the first
// one authenticates fine until that expires and then fails everywhere at
// once, with an error about an expired certificate and nothing pointing at
// the line that read it.
func (i *Identity) certificate() (*tls.Certificate, error) {
	info, err := os.Stat(i.cfg.CertFile)
	if err != nil {
		return nil, fmt.Errorf("stat the certificate: %w", err)
	}

	i.mu.RLock()
	cached, at := i.cert, i.loadedAt
	i.mu.RUnlock()

	if cached != nil && !info.ModTime().After(at) {
		return cached, nil
	}

	pair, err := tls.LoadX509KeyPair(i.cfg.CertFile, i.cfg.KeyFile)
	if err != nil {
		// A rotation is not atomic across two files, so a read that
		// catches it half-done fails here. The previous certificate is
		// still valid, so serving it beats refusing the connection.
		if cached != nil {
			return cached, nil
		}

		return nil, fmt.Errorf("load the certificate and its key: %w", err)
	}

	i.mu.Lock()
	i.cert, i.loadedAt = &pair, info.ModTime()
	i.mu.Unlock()

	return &pair, nil
}

// Server is the configuration for this service's own listener: it presents
// the mounted certificate, requires one from every caller, and admits only
// the accounts the configuration named.
func (i *Identity) Server() *tls.Config {
	if i == nil {
		return nil
	}

	return &tls.Config{
		MinVersion: tls.VersionTLS13,
		GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
			return i.certificate()
		},
		// The chain is verified by the library against the trust bundle;
		// WHO the verified peer is, is this package's question.
		ClientAuth:            tls.RequireAndVerifyClientCert,
		ClientCAs:             i.roots,
		VerifyPeerCertificate: i.verifyPeer,
	}
}

// Client is the configuration for connections this service makes: it
// presents the same certificate and checks the server's IDENTITY.
//
// It does not check the server's name, and `InsecureSkipVerify` is how that
// is expressed — which reads alarming and is not, because the chain and the
// identity are both verified below instead. The name is skipped because a
// platform's workload certificates carry no name: they carry an identity,
// which is a stronger statement. "I reached the address I meant to" is a
// weaker question than "the thing answering is the account I was told to
// trust", and the second one is the one being asked.
//
// The library's own verification has to be turned off to ask it, because it
// insists on the name. So this does the library's job by hand: build the
// chain against the trust bundle, then read the identity out of the leaf.
// Skipping either half would be the mistake the flag's name warns about.
func (i *Identity) Client() *tls.Config {
	if i == nil {
		return nil
	}

	return &tls.Config{
		MinVersion: tls.VersionTLS13,
		GetClientCertificate: func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
			return i.certificate()
		},
		// codeql[go/disabled-certificate-check] The certificate IS checked,
		// by verifyChainAndPeer below: it builds the chain against the trust
		// bundle and then reads the peer's identity. Only the NAME check is
		// skipped, because a platform's workload certificates carry no name.
		// Two tests hold the property this flag would otherwise destroy:
		// TestAClientRefusesAServerOutsideItsTrustBundle and
		// TestAClientRefusesAServerItWasNotToldToTrust.
		//nolint:gosec // see above
		InsecureSkipVerify:    true,
		VerifyPeerCertificate: i.verifyChainAndPeer,
	}
}

// verifyChainAndPeer does what the library would have done, plus the part it
// cannot: verify the chain against the trust bundle, then check the peer's
// identity. It is used where the name check had to be turned off.
func (i *Identity) verifyChainAndPeer(rawCerts [][]byte, _ [][]*x509.Certificate) error {
	if len(rawCerts) == 0 {
		return fmt.Errorf("the peer presented no certificate")
	}

	certs := make([]*x509.Certificate, 0, len(rawCerts))

	for _, raw := range rawCerts {
		cert, err := x509.ParseCertificate(raw)
		if err != nil {
			return fmt.Errorf("parse the peer's certificate: %w", err)
		}

		certs = append(certs, cert)
	}

	intermediates := x509.NewCertPool()
	for _, cert := range certs[1:] {
		intermediates.AddCert(cert)
	}

	chains, err := certs[0].Verify(x509.VerifyOptions{
		Roots:         i.roots,
		Intermediates: intermediates,
		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
	})
	if err != nil {
		return fmt.Errorf("the peer's certificate does not chain to the trust bundle: %w", err)
	}

	return i.verifyPeer(nil, chains)
}

// verifyPeer runs after the library has verified the chain. It reads the
// peer's identity out of the leaf and checks it against the allow-list.
func (i *Identity) verifyPeer(_ [][]byte, chains [][]*x509.Certificate) error {
	if len(chains) == 0 || len(chains[0]) == 0 {
		return fmt.Errorf("the peer presented no verified certificate")
	}

	peer, err := i.identityOf(chains[0][0])
	if err != nil {
		return err
	}

	for _, allowed := range i.cfg.Peers {
		if allowed == peer {
			return nil
		}
	}

	// Said HERE, because the caller will not be told. The refusal it
	// receives is a bare alert; this line is the only record that the
	// service made a decision rather than breaking.
	// Background, because there is none: this runs inside a TLS handshake,
	// before any request exists to carry one.
	i.log.WarnContext(context.Background(), "refused a peer",
		slog.String("namespace", peer.Namespace),
		slog.String("serviceAccount", peer.ServiceAccount))

	return fmt.Errorf("%s/%s is not a peer this service admits", peer.Namespace, peer.ServiceAccount)
}

// identityOf reads the account out of a certificate's identity URI.
//
// The shape is the SPIFFE one: spiffe://<trust domain>/ns/<namespace>/sa/<account>.
// Anything else is refused rather than guessed at, because a partial match
// here is an identity nobody meant to grant.
func (i *Identity) identityOf(leaf *x509.Certificate) (Peer, error) {
	for _, u := range leaf.URIs {
		if u.Scheme != "spiffe" {
			continue
		}

		if u.Host != i.cfg.TrustDomain {
			return Peer{}, fmt.Errorf("the peer's identity belongs to trust domain %q, not %q", u.Host, i.cfg.TrustDomain)
		}

		return parsePath(u)
	}

	return Peer{}, fmt.Errorf("the peer's certificate carries no identity")
}

func parsePath(u *url.URL) (Peer, error) {
	parts := strings.Split(strings.TrimPrefix(u.Path, "/"), "/")
	if len(parts) != 4 || parts[0] != "ns" || parts[2] != "sa" {
		return Peer{}, fmt.Errorf("the peer's identity %q does not name a namespace and an account", u.String())
	}

	if parts[1] == "" || parts[3] == "" {
		return Peer{}, fmt.Errorf("the peer's identity %q names an empty namespace or account", u.String())
	}

	return Peer{Namespace: parts[1], ServiceAccount: parts[3]}, nil
}

File: python/src/truvity_policy/transport.py
"""Turn the ``tls`` fragment of a configuration into the contexts a service needs.

It does three things, and deliberately no more:

1. loads the certificate the platform mounted, and RELOADS it when the
   platform replaces it;
2. presents it, as a client and as a server;
3. admits a peer by the ACCOUNT it runs as, read from the identity in its
   certificate, against a list the configuration gives.

It does not fetch a certificate, mint one, or talk to an authority. That is
the platform's job, and a service that did it would be asserting an identity
rather than presenting one it was given.

The Go package of the same name implements the same three things against the
same fragment. One difference is real and is not hidden here: Go can decide
whether to admit a peer DURING the handshake, and Python cannot. Python's
``ssl`` module has no verification callback, so the chain is verified by
OpenSSL and the identity is checked immediately afterwards, by the caller,
against the certificate the handshake already produced. A connection is
therefore refused a moment later than it would be in Go — after the
handshake completes, before anything is read or written — and
:func:`verify_peer` is the call that does it. Forgetting it leaves a service
that verifies a certificate chain and admits anybody who holds one, which is
the failure that looks like success from every other angle.
"""

from __future__ import annotations

import ssl
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING, Any, Literal
from urllib.parse import urlsplit

if TYPE_CHECKING:
    from collections.abc import Iterable

__all__ = ["Identity", "Peer", "TransportError", "load"]

Mode = Literal["off", "permissive", "strict"]

# The identity's shape: spiffe://<trust domain>/ns/<namespace>/sa/<account>.
_SCHEME = "spiffe"
_PATH_LENGTH = 4

# Stated rather than inherited, and the same floor the Go package sets.
#
# `create_default_context` already refuses anything below TLS 1.2, so this
# raises the floor by one version — but the reason to write it down is that
# the default is a property of the interpreter, and a service's transport
# floor should not change because a base image did. Both ends of every
# connection here are workloads this platform issued identities to, so there
# is nothing old to be compatible with.
_MINIMUM_VERSION = ssl.TLSVersion.TLSv1_3


class TransportError(Exception):
    """A configuration that cannot be loaded, or a peer that is not admitted.

    It names the identity it refused, because "TLS handshake failed" is the
    error people spend afternoons on. It never contains a key or anything
    read from one.
    """


@dataclass(frozen=True, slots=True)
class Peer:
    """An account. Never an address: an address resolves to whoever holds it today."""

    namespace: str
    service_account: str

    def __str__(self) -> str:
        """Render as the identity reads, so a log line can be searched for."""
        return f"{self.namespace}/{self.service_account}"


class Identity:
    """A loaded workload identity: what to present, and who to admit."""

    def __init__(  # noqa: PLR0913 (every one of them is a separate decision)
        self,
        *,
        mode: Mode,
        cert_file: str,
        key_file: str,
        ca_file: str,
        trust_domain: str,
        peers: Iterable[Peer] = (),
    ) -> None:
        """Hold the paths. Nothing is read until a context is asked for."""
        self.mode: Mode = mode
        self._cert_file = cert_file
        self._key_file = key_file
        self._ca_file = ca_file
        self._trust_domain = trust_domain
        self._peers = frozenset(peers)

    @property
    def peers(self) -> frozenset[Peer]:
        """Who may call. Empty admits nobody, which is the right default."""
        return self._peers

    def client_context(self) -> ssl.SSLContext:
        """Build a context for connections this service MAKES.

        Built fresh on every call, and that is the point. A platform rotates
        these on an hour or less; a process that held the context it built at
        start-up authenticates fine until the certificate expires and then
        fails everywhere at once, with an error about expiry and nothing
        pointing at the line that read it. Building per connection is what
        Python makes natural, and it is why a client needs no reload
        machinery at all.

        Host name checking is OFF, and that is a decision rather than an
        oversight: a platform's workload certificate carries no name. It
        carries an identity, which is the stronger statement. "I reached the
        address I meant to" is a weaker question than "the thing answering is
        the account I was told to trust", and the chain is still verified.
        The second question is asked by :meth:`verify_peer`, which the caller
        MUST call once the connection is up.
        """
        context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH, cafile=self._ca_file)
        context.minimum_version = _MINIMUM_VERSION
        context.check_hostname = False
        context.verify_mode = ssl.CERT_REQUIRED
        context.load_cert_chain(self._cert_file, self._key_file)
        return context

    def server_context(self) -> ssl.SSLContext:
        """Build a context for this service's own listener.

        A certificate is REQUIRED from every caller, so a connection without
        one never reaches the application. Which accounts are admitted is
        still :meth:`verify_peer`'s question, asked once the handshake is
        done.

        A server built on this reads its certificate when the context is
        built. Rotation is therefore the awkward case in Python and is not
        solved here: a process-per-worker server recycles its workers within
        the certificate's lifetime, which is ordinary and needs no new
        machinery, and anything else sits behind the terminating proxy. See
        docs/canon/python.md.
        """
        context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH, cafile=self._ca_file)
        context.minimum_version = _MINIMUM_VERSION
        context.verify_mode = ssl.CERT_REQUIRED
        context.load_cert_chain(self._cert_file, self._key_file)
        return context

    def peer_of(self, certificate: dict[str, Any] | None) -> Peer:
        """Read the account out of a peer certificate's identity.

        Takes what ``SSLSocket.getpeercert()`` returns. Anything that is not
        the expected shape is refused rather than guessed at, because a
        partial match here is an identity nobody meant to grant.
        """
        if not certificate:
            raise TransportError("the peer presented no certificate")

        uris = [
            value
            for kind, value in certificate.get("subjectAltName", ())
            if kind == "URI" and value.startswith(f"{_SCHEME}://")
        ]
        if not uris:
            raise TransportError("the peer's certificate carries no workload identity")
        if len(uris) > 1:
            # Two identities in one certificate is not a peer to guess about.
            raise TransportError(f"the peer's certificate carries {len(uris)} identities")

        parsed = urlsplit(uris[0])
        if self._trust_domain and parsed.netloc != self._trust_domain:
            raise TransportError(
                f"identity {uris[0]} is from trust domain {parsed.netloc!r}, "
                f"not {self._trust_domain!r}",
            )

        parts = parsed.path.strip("/").split("/")
        if len(parts) != _PATH_LENGTH or parts[0] != "ns" or parts[2] != "sa":
            raise TransportError(f"identity {uris[0]} is not in the shape this service reads")
        return Peer(namespace=parts[1], service_account=parts[3])

    def verify_peer(self, certificate: dict[str, Any] | None) -> Peer:
        """Admit the peer, or raise saying which account was refused.

        Call this once a connection is up and before anything is read or
        written on it. The chain is already verified by then — OpenSSL did
        that during the handshake — and this is the other half: WHO the
        verified certificate belongs to.
        """
        peer = self.peer_of(certificate)
        if peer not in self._peers:
            raise TransportError(f"refused a peer: {peer} is not on this service's list")
        return peer


def load(cfg: dict[str, Any] | None) -> Identity | None:
    """Read the ``tls`` fragment, or return None when there is nothing to load.

    None means "serve cleartext", not an error. `off` is the default and
    always will be: a chart is installable by someone whose platform
    provides none of this, and a default that assumed one would produce a
    service waiting forever for a volume nobody serves.
    """
    if not cfg:
        return None
    mode = cfg.get("mode", "off")
    if mode == "off":
        return None
    if mode not in ("permissive", "strict"):
        raise TransportError(f"tls.mode {mode!r} is not one this service knows")

    missing = [key for key in ("certFile", "keyFile", "caFile") if not cfg.get(key)]
    if missing:
        raise TransportError(
            f"tls.mode is {mode} but {', '.join(missing)} "
            f"{'is' if len(missing) == 1 else 'are'} not set: "
            "the platform mounts the identity, and this is where it says where",
        )
    for key in ("certFile", "keyFile", "caFile"):
        if not Path(cfg[key]).is_file():
            raise TransportError(f"tls.{key} is {cfg[key]!r}, which is not a file")
    if not cfg.get("trustDomain"):
        # Without it a peer from ANY trust domain is admitted by account
        # name alone, which is a different service's `default` looking
        # exactly like this one's.
        raise TransportError(
            "tls.trustDomain is required once tls.mode is not off: "
            "without it a peer from any trust domain is admitted",
        )

    return Identity(
        mode=mode,
        cert_file=cfg["certFile"],
        key_file=cfg["keyFile"],
        ca_file=cfg["caFile"],
        trust_domain=cfg["trustDomain"],
        peers=[
            Peer(namespace=p["namespace"], service_account=p["serviceAccount"])
            for p in cfg.get("peers", [])
        ],
    )
